Assets/RGLUnityPlugin/Scripts/LowLevelWrappers/RGLGraph.cs

```diff
@@ -0,0 +1,333 @@
+        public int GetResultData<T>(ref T[] data) where T : unmanaged
+        {
+            ValidateOutputNode();
+            unsafe
+            {
+                return RGLNativeAPI.GraphGetResult<T>(nodes.Last().Node, outputField, ref data, sizeof(T));
+            }
+        }
+
+        public int GetResultDataRaw(ref byte[] data, int expectedPointSize)
+        {
+            ValidateOutputNode();
+            return RGLNativeAPI.GraphGetResult<byte>(nodes.Last().Node, outputField, ref data, expectedPointSize);
+        }
+
+        public void Run()
+        {
+            if (nodes.Count == 0)
+            {
+                throw new RGLException("Attempted to run empty graph!");
+            }
+            RGLNativeAPI.GraphRun(nodes.First().Node);
+        }
+
+        public void Destroy()
+        {
+            DisconnectAllChildGraphs();
+            DisconnectAllParentGraphs();
+            if (nodes.Count > 0)
+            {
+                RGLNativeAPI.GraphDestroy(nodes.First().Node);
+                nodes.Clear();
+            }
+        }
+
+        //// PRIVATE HELPERS ////
+        private void ValidateNewNode(string identifier)
+        {
+            var nodeFilter = nodes.Where(n => n.Identifier == identifier);
+            if (nodeFilter.Count() != 0)
+            {
+                throw new RGLException($"Attempted to add node '{identifier}' twice!");
+            }
+        }
+
+        private void ValidateExistingNode(string identifier, RGLNodeType desiredType)
+        {
+            var nodeFilter = nodes.Where(n => n.Identifier == identifier);
+            if (nodeFilter.Count() != 1)
+            {
+                throw new RGLException($"Attempted to access node '{identifier}' but it was not found!");
+            }
+            if (nodeFilter.First().Type != desiredType)
+            {
+                throw new RGLException($"Attempted to access node '{identifier}' but node type mismatch!");
+            }
+        }
+
+        private void ValidateOutputNode()
+        {
+            if (outputField == RGLField.UNKNOWN)
+            {
+                throw new RGLException("Attempted to get result data but format node was not found!");
+            }
+            RGLNodeType lastNodeType = nodes.Last().Type;
+            if (!(lastNodeType == RGLNodeType.POINTS_FORMAT || lastNodeType == RGLNodeType.POINTS_YIELD))
+            {
+                throw new RGLException("Attempted to get result data but format node is not the last node in the graph!");
+            }
+        }
+
+        private void AddNode(RGLNodeHandle nodeHandle)
+        {
+            if (nodes.Count == 0)
+            {
+                nodes.Add(nodeHandle);
+                return;
+            }
+
+            if (childGraphs.Count > 0)
+            {
+                Debug.LogWarning("Added node to already connected graph. Removing child graphs...");
+                DisconnectAllChildGraphs();
+            }
```

Comment:
Disconnecting all child graphs will make them return constant output on `rgl_graph_run(...)`, which may be surprising if the warning goes unnoticed. An error is harder to omit usually. At the moment, I don't see any drawbacks to throwing an exception. What do you think?

Reply:
We had an offline discussion about graph behavior when adding a new node between already connected graphs. We decided to not support these operations at the moment. I agree that exception is better than log message and unpredicted behavior.